File: project09/java/Temperature.java
/* Temperature.java implements class
 *
 *
 * Haim Hong
 * March 17 2022
*/
import java.util.Scanner;
import java.lang.Character;

class Temperature {

    private Double myDegrees;
    private char myScale;
    private static final char kelvin = 'K';
    private static final char celsius = 'C';
    private static final char fahrenheit = 'F';

    // Temperature contructor with initialization
    public Temperature(Double deg, char sca){
        myDegrees = deg;
        myScale = Character.toUpperCase(sca);
        if (!checkValid(myDegrees, myScale)){
            throw new IllegalArgumentException("Invalid scale");
        }
    }

    // extractor for degree of a temp
    public final Double getDegrees(){
        return myDegrees;
    }

    // extractor for scale of a temp
    public final Character getScale(){
        return myScale;
    }

    // converts any to fahrenheit if its not if it is just return itself
    // calculates and use that to return new temperature 
    public Temperature toFahrenheit(){
        if (myScale == kelvin){
            return new Temperature((myDegrees - 273.15) * (9.0/5.0) + 32.0, fahrenheit);
        }
        else if (myScale == celsius){
            return new Temperature((myDegrees * (9.0/5.0)) + 32.0, fahrenheit);
        }
        else {
            return this;
        }
    }

    // converts any to celsius if its not if it is just return itself
    // calculates and use that to return new temperature 
    public Temperature toCelsius(){
        if (myScale == kelvin){
            return new Temperature(myDegrees - 273.15, celsius);
        }
        else if (myScale == fahrenheit){
            return new Temperature((myDegrees - 32.0) * (5.0/9.0), celsius);
        }
        else {
            return this;
        }
    }

    // converts any to kelvin if its not if it is just return itself
    // calculates and use that to return new temperature 
    public Temperature toKelvin(){
        if (myScale == fahrenheit){
            return new Temperature((myDegrees - 32.0) * (5.0/9.0) + 273.15, kelvin);
        }
        else if (myScale == celsius){
            return new Temperature(myDegrees + 273.15, kelvin);
        }
        else {
            return this;
        }
    }

    // reads the temperature from user input
    // returns temperature with user inputted degree and scale
    public static Temperature readTemperature(Scanner input){

        String temps = input.nextLine();
        String [] splite = temps.split(" ");
        Character scale = splite[1].charAt(0);
        Double degree = Double.parseDouble(splite[0]);
        try {
            Double degree1 = Double.parseDouble(splite[0]);
        } catch (NumberFormatException A){
            return null;
        } if (!checkValid(degree, scale)){
            return null;
        }
        return new Temperature(degree, scale);
    }

    // prints out the tempeerature 
    public final void printTemperature(){
        System.out.printf("%.2f\t%c",myDegrees, myScale);
    }

    // raises the value of temperature 
    public Temperature raise(Double moreDegree){
        return new Temperature(myDegrees + moreDegree, myScale);
    }

    // lowers the value of temperature 
    public Temperature lower(Double moreDegree, char scale){
        if (!checkValid(moreDegree, scale)){
            return null;
        }
        return new Temperature((myDegrees - moreDegree), myScale);
    }
    
    // checks if the temperature are same
    // returns true or false
    public boolean equals(Temperature cTemp){
        //System.out.println("==");
        //System.out.println(myDegrees);
        //System.out.println(compareT(cTemp).getDegrees());
        //System.out.println(myDegrees - compareT(cTemp).getDegrees() == 0.0);
        return  myDegrees - compareT(cTemp).getDegrees() == 0.0;
    }

    // checks if the temperature is lower than other temp
    // returns true or false
    public boolean lowerThan(Temperature cTemp){
        return  myDegrees < compareT(cTemp).getDegrees() ;
    }

    // compares the two temperatures and converts it to be same
    // returns the temp that is converted
    public Temperature compareT(Temperature cTemp){
        if (myScale != cTemp.getScale()){
            if (myScale == celsius){
                return cTemp.toCelsius();
            }
            else if (myScale == fahrenheit){
                return cTemp.toFahrenheit();
            }
            else {
                return cTemp.toKelvin();
            }
        }
        return cTemp;
    }
    // check if the temperatures are valid
    // returns true if all all correct else false
    private static boolean checkValid(Double degree, char scale){
        if (Character.isLowerCase(scale)){
            Character.toUpperCase(scale);
        }
        if ((scale == kelvin && degree >= 0) || (scale == celsius && degree >= -273.15) || (scale == fahrenheit && degree >= -459.67)){
            return true;
        }
        return false;
    }
}

File: project09/clojure/src/temperature.clj
;;;; module for temperature type abstration
;;;; 
;;;; Haim Hong
;;;; March 18 2022
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; checkValid checks if the scale it valid
;; Receive the scale
;; returns true or falase
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn checkValid [degree scale]
  (if (or (and (= scale "C") (>= degree -273.15)) (and (= scale "F")(>= degree -459.67)) (and (= scale "K")(>= degree 0.0))) ;;)
    true
    false)
)

;; create a record type named Temperature
(defrecord Temperature [myDegree myScale])

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; make-Temp constructs a Temperature object from three strings. 
;; Receive: myDegree my
;; Return: the Temp myDeg mySca
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn make-Temp [myDeg mySca]
  (if (checkValid myDeg mySca)
    (->Temperature myDeg mySca)
    false))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; getDegree gets the degree from Temp
;; receive Temp 
;; Return: myDegree
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn getDegree [^Temperature aTemp]
  (:myDegree aTemp))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; getDegree gets the degree from Temp
;; receive Temp 
;; Return: myScale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn getScale [^Temperature aTemp]
  (:myScale aTemp))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toFahrenheit changes the degree to Fahrenheit
;; receive Temp 
;; Return: new Temp with changes degree and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toFahrenheit [^Temperature aTemp]
  (cond
  (= (getScale aTemp) "C") (make-Temp (+ (* (getDegree aTemp) (/ 9.0 5.0)) 32.0) "F")
  (= (getScale aTemp) "K") (make-Temp (+ (* (- (getDegree aTemp) 273.15) (/ 9.0 5.0)) 32.0) "F")
  (= (getScale aTemp) "F") (make-Temp (getDegree aTemp) "F")))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toCelsius changes the degree to Celsius
;; receive Temp 
;; Return: new Temp with changes degree and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toCelsius [^Temperature aTemp]
  (cond
    (= (getScale aTemp) "F") (make-Temp (/ (* (- (getDegree aTemp) 32.0) 5.0) 9.0) "C")
    (= (getScale aTemp) "K") (make-Temp (- (getDegree aTemp) 273.15) "C")
    (= (getScale aTemp) "C") (make-Temp (getDegree aTemp) "C")))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; toKelvin changes the degree to Kelvin
;; receive Temp 
;; Return: new Temp with changes degree and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toKelvin [^Temperature aTemp]
  (cond
    (= (getScale aTemp) "C") (make-Temp (+ (getDegree aTemp) 273.15) "K")
    (= (getScale aTemp) "F") (make-Temp (* (+ (getDegree aTemp) 459.67) (/ 5 9)) "K")
    (= (getScale aTemp) "K") (make-Temp (getDegree aTemp) "K"))
    )
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; raise adds the temperature wiht the given temp
;; receive Temp 
;; Return: returnes temperature with new temp and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn raise [^Temperature aTemp aDeg]
  (make-Temp (+ (getDegree aTemp) aDeg) (getScale aTemp)))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; lower subtracts from the temperature wiht the given temp
;; receive Temp 
;; Return: returnes temperature with new temp and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn lower [^Temperature aTemp aDeg]
  (make-Temp (- (getDegree aTemp) aDeg) (getScale aTemp)))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; readTemp reads the Temp
;; receive Temp 
;; Return: new Temp with user input of DEgree and Scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn readTemp []
  (make-Temp (read-string (read-line)) (read-line)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; toString() returns string of temp and scale
;;; Receive Temp
;;; Return: temperature to decimal points of 2 and scale of it seperated by space
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn toString [^Temperature aTemp]
  (str (format "%.2f" (double (getDegree aTemp))) " " (getScale aTemp)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; printTemp prints the temperature out
;; receive Temp 
;; Return: prints out the temp with scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn printTemp [^Temperature aTemp]
  (print (toString aTemp)))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; equals checkes between two temperatures if they are equal by convetting it to fahrenheit
;; receive Temp 
;; Return: returnes true or false
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn equals [^Temperature aTemp ^Temperature aTemp2]
  (if (= (getScale aTemp) (getScale aTemp2))
      (= (getDegree aTemp) (getDegree aTemp2))
      (= (getDegree (toFahrenheit aTemp)) (getDegree (toFahrenheit aTemp2)))
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; lessThan checkes if temp is lower than other temp by converting them to fahrenheit
;; receive Temp 
;; Return: returnes true or false
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn lessThan [^Temperature aTemp ^Temperature aTemp2]
  (if (= (getScale aTemp) (getScale aTemp2))
    (< (getDegree aTemp) (getDegree aTemp2))
    (< (getDegree (toFahrenheit aTemp)) (getDegree (toFahrenheit aTemp2)))
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; lessThan checkes if temp is lower than other temp by converting them to fahrenheit
;; receive Temp 
;; Return: returnes true or false
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn printTable [baseT limitT stepV]

  (loop [base baseT
         limit limitT]
        (if (lessThan base limit)
          (do
            (printTemp (toFahrenheit base))
            (print "    ")
            (printTemp (toCelsius base))
            (print "    ")
            (printTemp (toKelvin base))
            (print "\n")
            (recur (raise base stepV) limit))
          )))